packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -288,48 +277,192 @@ class CupertinoDialogAction extends StatelessWidget {
+            textAlign: TextAlign.center,
+            child: message,
+          ),
+        ),
+      );
+    }
+
+    if (titleContentGroup.isEmpty) {
+      return new SingleChildScrollView(
+        controller: scrollController,
+        child: new Container(width: 0.0, height: 0.0),
+      );
+    }
+
+    // Add padding between the widgets if necessary.
+    if (titleContentGroup.length > 1) {
+      titleContentGroup.insert(1, const Padding(padding: const EdgeInsets.only(top: 8.0)));
+    }
+
+    return new Flexible(
+      flex: 3,
+      child: new CupertinoScrollbar(
+        child: new SingleChildScrollView(
+          controller: scrollController,
+          child: new Column(
+            mainAxisSize: MainAxisSize.max,
+            crossAxisAlignment: CrossAxisAlignment.stretch,
+            children: titleContentGroup,
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+
+/// An Action Items section typically used in a [CupertinoAlertDialog].
+///
+/// If _shouldLayoutActionsVertically(_) is true, they are laid out vertically
+/// in a column; else they are laid out horizontally in a row. If there isn't
+/// enough room to show all the [children] vertically, they are wrapped in a
+/// [CupertinoScrollbar] widget. If [children] is null or empty, it returns null.
+class _CupertinoAlertActionSection extends StatelessWidget {
+  const _CupertinoAlertActionSection({
     Key key,
     this.children,
+    this.scrollController,
   }) : super(key: key);
 
   final List<Widget> children;
 
+  /// A scroll controller that can be used to control the scrolling of the
+  /// message in the dialog.
+  ///
+  /// Defaults to null, and is typically not needed, since most alert messages
+  /// are short.
+  final ScrollController scrollController;
+
   @override
   Widget build(BuildContext context) {
+    if (children == null || children.isEmpty) {
+      return new SingleChildScrollView(
+        controller: scrollController,
+        child: new Container(width: 0.0, height: 0.0),
+      );
+    }
+
     final List<Widget> buttons = <Widget>[];
 
-    for (Widget child in children) {
-      // TODO(abarth): Listen for the buttons being highlighted.
-      // TODO(gspencer): These buttons don't lay out in the same way as iOS.
-      // When they get wide, they should stack vertically instead of in a row.
-      // When they get really big, the vertically-stacked buttons should scroll
-      // (separately from the top message).
-      buttons.add(new Expanded(child: child));
+    Widget getActionWidgetForIndex(int index) {
+      if (_shouldLayoutActionsVertically(children.length)) {
+        // Skip the first divider for vertical layout.
+        if (index==0) {
+          return children[index];
+        }
+        return new CustomPaint(
+          painter: new _CupertinoVerticalDividerPainter(),
+          child: children[index],
+        );
+      }
+      return new Expanded(child: children[index]);
```

Comment:
as before, returning an Expanded is generally an anti-pattern. prefer to move that to the callsite.